WalletWasabi/WabiSabi/Backend/PostRequests/InputRegistrationHandler.cs

```diff
@@ -140,20 +140,20 @@ public static class InputRegistrationHandler
 				commitWeightCredentialResponse.Commit());
 		}
 
-		private static void RemoveDuplicateAlices(IDictionary<Guid, Round> rounds, Alice alice)
+		private static void RemoveDuplicateAlices(IDictionary<Guid, Round> roundsWithId, Alice alice)
 		{
-			foreach (var (otherRound, op) in rounds
-							.Values
-							.SelectMany(otherRound => alice
-								.Coins
-								.Select(x => x.Outpoint)
-								.Select(op => (otherRound, op))))
+			var rounds = roundsWithId.Values;
+			if (rounds.Any(x => x.Phase != Phase.InputRegistration))
 			{
-				if (otherRound.Phase != Phase.InputRegistration)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceAlreadyRegistered);
-				}
-				if (otherRound.Alices.RemoveAll(x => x.Coins.Select(x => x.Outpoint).Contains(op)) > 0)
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceAlreadyRegistered);
+			}
+			
+			var aliceOutPoints = alice.Coins.Select(x => x.Outpoint).ToHashSet();
+			var flattenTable = rounds.SelectMany(x => x.Alices.SelectMany(y => y.Coins.Select(z => (Round: x, Alice: y, Output: z.Outpoint))));
+
+			foreach (var (round, aliceInRound, _) in flattenTable.Where(x => aliceOutPoints.Contains(x.Output)).ToArray())
```

Comment:
This is not about performance, as it probably matter since these collections are restricted to several hundred elements, but it seems a bit counter intuitive to me, I would have expected just a map keyed by outpoint instead of a list of tuples and then what is effectively a left join. Maybe it makes sense for `Arena` to just maintain an index of all `Alice`s and the rounds they are in, keyed by outpoint?

Reply:
Yes, it make sense. Once `Round` implements `AddAlice` then it can keep the map. In the meanwhile I think the solution in this PR is better that the previous one.